orangecontrib/spectroscopy/widgets/owsnr.py

```diff
@@ -0,0 +1,192 @@
+import numpy as np
+
+import Orange.data
+from Orange.data.filter import SameValue, FilterDiscrete, Values
+from Orange.widgets.widget import OWWidget, Msg, Input, Output
+from Orange.widgets import gui, settings
+from Orange.widgets.utils.itemmodels import DomainModel
+
+
+class OWSNR(OWWidget):
+    # Widget's name as displayed in the canvas
+    name = "snr"
+
+    # Short widget description
+    description = (
+        "Calculates SNR, averages or Standard Deviation.")
+
+    icon = "icons/snr.svg"
+
+    # Define inputs and outputs
+    class Inputs:
+        data = Input("Data", Orange.data.Table, default=True)
+
+    class Outputs:
+        final_data = Output("SNR", Orange.data.Table, default=True)
+
+    OUT_OPTIONS = {'Signal-to-noise ratio': 0, #snr
+                   'Average': 1, # average
+                   'Standard Deviation': 2} # std
+
+    settingsHandler = settings.DomainContextHandler()
+    group_x = settings.ContextSetting(None)
+    group_y = settings.ContextSetting(None)
+    group = settings.ContextSetting(None)
+    out_choiced = settings.ContextSetting(0)
+
```

Comment:
Context settings are those that depend on the structure of Table's domain. They mainly exist to prevent incompatible settings. They are, obviously, Group X and Group Y, but not that many others. If you do not want the other settings to be saved as defaults, you can add the `schema_only` flag.